libs/ml/include/ml/dataloaders/commodity_dataloader.hpp

```diff
@@ -0,0 +1,241 @@
+ * By default this skips the first row amd columns as these are assumed to be indices.
+ * @param xfilename Name of the csv file with inputs
+ * @param yfilename name of the csv file with labels
+ */
+template <typename LabelType, typename InputType>
+void CommodityDataLoader<LabelType, InputType>::AddData(std::string const &xfilename,
+                                                        std::string const &yfilename)
+{
+  std::pair<SizeType, SizeType> xshape = count_rows_cols(xfilename);
+  std::pair<SizeType, SizeType> yshape = count_rows_cols(yfilename);
+  SizeType                      row    = xshape.first;
+  SizeType                      col    = xshape.second;
+
+  data_.Reshape({row - rows_to_skip_, col - cols_to_skip_});
+  labels_.Reshape({yshape.first - rows_to_skip_, yshape.second - cols_to_skip_});
+  assert(xshape.first == yshape.first);
+  // save the number of rows in the data
+  size_ = row - rows_to_skip_;
+
+  // read csv data into buffer_ array
+  std::ifstream file(xfilename);
+  std::string   buf;
+  char          delimiter = ',';
+  std::string   field_value;
+
+  for (SizeType i = 0; i < rows_to_skip_; i++)
+  {
+    std::getline(file, buf, '\n');
+  }
+
+  row = 0;
+  while (std::getline(file, buf, '\n'))
+  {
+    col = 0;
+    std::stringstream ss(buf);
+    for (SizeType i = 0; i < cols_to_skip_; i++)
+    {
+      std::getline(ss, field_value, delimiter);
+    }
+
+    while (std::getline(ss, field_value, delimiter))
+    {
+      data_(row, col) = static_cast<DataType>(stod(field_value));
+      ++col;
+    }
+    ++row;
+  }
+
+  file.close();
+
+  file.open(yfilename);
+  for (SizeType i = 0; i < rows_to_skip_; i++)
+  {
+    std::getline(file, buf, '\n');
+  }
+
+  row = 0;
+  while (std::getline(file, buf, '\n'))
+  {
+    col = 0;
+    std::stringstream ss(buf);
+    for (SizeType i = 0; i < cols_to_skip_; i++)
+    {
+      std::getline(ss, field_value, delimiter);
+    }
+
+    while (std::getline(ss, field_value, delimiter))
+    {
+      labels_(row, col) = static_cast<DataType>(stod(field_value));
+      ++col;
+    }
+    ++row;
+  }
+}
+
+/**
+ * Gets the next pair of data and labels
+ * @tparam LabelType type for the labels
+ * @tparam InputType type for the data
+ * @return Pair of input and output
+ */
+template <typename LabelType, typename InputType>
+typename CommodityDataLoader<LabelType, InputType>::ReturnType
+CommodityDataLoader<LabelType, InputType>::GetNext()
+{
+  if (this->random_mode_)
+  {
+    GetAtIndex(static_cast<SizeType>(rand.generator()) % Size());
+    return buffer_;
+  }
+  else
+  {
+    GetAtIndex(static_cast<SizeType>(cursor_++));
+    return buffer_;
+  }
+}
+
+/**
+ * Constructor for CommodityDataLoader
+ * @tparam LabelType Type for the labels, usually tensor<float>
+ * @tparam InputType Type for the data, usually tensor<float>
+ * @param random_mode whether to access the data randomly
+ */
+template <typename LabelType, typename InputType>
```

Comment:
when the implementation is just an empty brace you can just do that inside the class